march_hardware/src/Joint.cpp

```diff
@@ -65,34 +70,94 @@ void Joint::actuateRad(double target_position)
   this->imc_->actuateRad(target_position);
 }
 
-double Joint::getAngleRadAbsolute()
+void Joint::readEncoders(const ros::Duration& elapsed_time)
 {
   if (!this->hasIMotionCube())
   {
     ROS_WARN("[%s] Has no iMotionCube", this->name_.c_str());
-    return -1;
+    return;
   }
-  return this->imc_->getAngleRadAbsolute();
-}
 
-double Joint::getAngleRadIncremental()
-{
-  if (!this->hasIMotionCube())
+  if (this->receivedDataUpdate())
   {
-    ROS_WARN("[%s] Has no iMotionCube", this->name_.c_str());
-    return -1;
+    const double new_incremental_position = this->imc_->getAngleRadIncremental();
+    const double new_absolute_position = this->imc_->getAngleRadAbsolute();
+
+    // Get velocity from encoder position
+    const double incremental_velocity = (new_incremental_position - this->incremental_position_) / elapsed_time.toSec();
+    const double absolute_velocity = (new_absolute_position - this->absolute_position_) / elapsed_time.toSec();
+
+    // If the difference in movement of the two encoders is larger than twice the sum of their precisions,
+    // one of them is probably an outlier.
+    if (std::abs(incremental_velocity - absolute_velocity) * elapsed_time.toSec() >
+        2 * (this->imc_->getAbsoluteRadPerBit() + this->imc_->getIncrementalRadPerBit()))
+    {
+      // Take the velocity that is closest to that of the previous timestep.
+      if (std::abs(incremental_velocity - this->velocity_) < std::abs(absolute_velocity - this->velocity_))
+      {
+        ROS_WARN("There was an outlier in the absolute encoder; old value: %f",
+                 this->absolute_position_, new_absolute_position);
+        this->velocity_ = incremental_velocity;
+      }
+      else
+      {
+        ROS_WARN("There was an outlier in the incremental encoder; old value: %f, new value: %f",
+                 this->incremental_position_, new_incremental_position);
+        this->velocity_ = absolute_velocity;
+      }
+    }
+    else
+    {
+      // Recalibrate joint position if it has drifted
+      if (std::abs(this->absolute_position_ - this->position_) > 2 * this->imc_->getAbsoluteRadPerBit())
+      {
+        this->position_ = this->absolute_position_;
+        ROS_INFO("Recalibrated joint position %s", this->name_.c_str());
+      }
+
+      // Take the velocity with the highest resolution.
+      if (this->imc_->getIncrementalRadPerBit() < this->imc_->getAbsoluteRadPerBit())
+      {
+        this->velocity_ = incremental_velocity;
+      }
+      else
+      {
+        this->velocity_ = absolute_velocity;
+      }
+    }
+
+    // Update position with the most accurate velocity
+    this->position_ += this->velocity_ * elapsed_time.toSec();
+    this->incremental_position_ = new_incremental_position;
+    this->absolute_position_ = new_absolute_position;
+  }
+  else
+  {
+    // Update positions with velocity from last time step
+    this->position_ += this->velocity_ * elapsed_time.toSec();
```

Comment:
Hmm good point. It is probably not that big a problem since you regularly re-calibrate anyway, but I changed it to use displacement instead of velocity.